### DISCUSSION TITLE
Better Interaction of Tasmota with RF chip EFM8BB1 in Sonoff RFR3 - start remote button pairing by tasmota command

### DISCUSSION THREAD

**Discussion by zhladik** [arendst/Tasmota#13160] [Q&A] [answered]
As I see in schema RF chip EFM8BB1  IN/OUT P13 is connected in parallel to Button and used to get programming 3s command and also for switching relay button signal. By this trick, there is only one GPIO0 interconnection, but there is not possible to start
RF remote button/switch pairing another way then 3s real button pressing. 

But in many real installations for example, on the ceiling, it is hard to reprogram the pairing of RF remote. It will be nice to get a command for 3s or 5s lowering (after changing the mode from input to output). No other mod is necessary. 

Another solution is to interconnect  P13 with some spare ESP GPIO for this. But it needs HW mod of all RFR3. 

It is also possible to interconnect serial RX/TX, but I don't know if default EFM8BB1 sends any info to serial. Maybe FW is the same as in Sonoff RF bridge?

**Comment by barbudor** [comment, marked as answer]
If the device is in the ceiling, you don't need the button. So feel free to set the GPIO0 as a Relay and control with POWER.
With PulseTime you can even define exactly the duration of the pulse

**Reply by zhladik** [reply]
Great, but probably usable only in an emergency - we can change template (reboot mandatory?), enter the command for 3s, pair remote button and then change template back to the input (and reboot again?). Without reverting template back will be a signal from EFM8BB1 after receiving RF button probably ignored...
Also, there is a question if the output of an ESP is shortage resistant. (EFM8BB1 probably is OK in open collector mode?)

**Reply by barbudor** [reply]
Why would you change the template dynamically ?
I don't know the schematic of the board and if you have it, please share it
But from what you say, the physical button is connected to an GPIO of the ESP and to what is also an input of the EFM.
So the pin of the EFM is not an output to the ESP

In your case the device is in the ceiling, and no-one will ever push the button, so forget about shortage on the ESP pin.
Just change it to Relay, use Pulsetime and that's it.

**Reply by zhladik** [reply]
It is necessary to enable again button function of Tasmota because the button press is emulated on RF code receive event by EFM8BB1.  Without this, we will lose RF functionality...

You can check in schematics that this crazy trick of piggybacking rf circuit to GPIO0 is used, EFM8BB1. Here is [Sonoff Schematics](https://wiki.iteadstudio.com/File:Sonoff-Schmatic.pdf)

This trick allows to use RF without ewelink fw modification (and also basic Tasmota works), but you will be not able to see received code, even distinguish between button press and rf event is impossible.

There is the missing circuit of 433 radio on this schematic - SYN470. By connecting the output of the radio signal detector to free GPIO, there is possible to get raw RF data, but I have not yet tested it yet. And make wire soldering on this small board is not too easy...

I will go to test template hack - it is possible to create a simple script with wget/culr GET requests to automate it. Reassigning of remote buttons is rare, so making it out of HA is acceptable....

**Reply by barbudor** [reply]
> It is necessary to enable again button function of Tasmota because the button press is emulated on RF code receive event by EFM8BB1. Without this, we will lose RF functionality...

From where do you get that ?
I see nothing in Tasmota code for module RFBridge that is using Button_1.
AFAIK, the communication between the EFM and the ESP is only through serial. There is no case where the EFM will assert the KEY line expecting the ESP to react.

**Reply by zhladik** [reply]
My explorations is about Tasmota basic RF R3. See [Itead RFR3](https://itead.cc/product/sonoff-rfr3/) (RF R2 and R1 use probably same interconnects). RF Bridge is probably a totally different beast. I will receive one next week. And I found no schematics. With HW hack and Sonoff-sensors Tasmota FW variant, there is probably possible to use close to half of Sonoff bridge functionality even on RFR3...

**Reply by barbudor** [reply]
ok, sorry for mistaking RFR3 and Bridge
In that case, yes you need to revert the template
The best would be an hardware hack driving the key signal low through a transistor from another free GPIO defined as Relay 2
